### PR TITLE
Fix issues count bug

### LINKED ISSUE #19349: Pull Request counter offset after deleting branch
### Description

possibly related to #19190

Unfortunately not able to reproduce.

Multiple PRs were based on a branch, which also had a PR to the master branch. upon closing the PR to the master branch and deleting the underlying branch, the other PRs were automatically closed. so far so good, but the pull request counter is not permanently off by 1. 

See screenshots.

Right now there are two PRs open but the counter says 2. If no pull requests are open, it says `-1` which is extra fun

### Gitea Version

1.15.11

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

image

### Git Version

_No response_

### Operating System

alpine

### How are you running Gitea?

from openRC

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by lunny**
Should fix #19349 , #19505

### CODE REVIEW

File: models/repo/repo.go [hunk 1]
```diff
@@ -765,35 +765,35 @@ func CountRepositories(ctx context.Context, opts CountRepositoryOptions) (int64,
...
-		}
-		if isClosed {
-			if _, err := e.ID(repoID).Decr("num_closed_pulls").Update(new(Repository)); err != nil {
-				return err
-			}
+	var field string
+	if isClosed {
+		field = "num_closed_issues"
+		if isPull {
+			field = "num_closed_pulls"
```
Comment (delvh) [inline models/repo/repo.go]:
```suggestion
	field := "num_"
	if isClosed {
		field += "closed_"
	}
	if isPull {
		field += "pulls"
	} else {
		field += "issues"
	}
```

File: models/repo/repo.go [hunk 2]
```diff
@@ -765,35 +765,35 @@ func CountRepositories(ctx context.Context, opts CountRepositoryOptions) (int64,
...
 		}
 	} else {
```
Comment (delvh) [inline models/repo/repo.go]:
```suggestion
```

File: models/repo/repo.go [hunk 3]
```diff
@@ -765,35 +765,35 @@ func CountRepositories(ctx context.Context, opts CountRepositoryOptions) (int64,
...
-		}
-		if isClosed {
-			if _, err := e.ID(repoID).Decr("num_closed_issues").Update(new(Repository)); err != nil {
-				return err
-			}
+		field = "num_issues"
+		if isPull {
+			field = "num_pulls"
 		}
 	}
```
Comment (delvh) [inline models/repo/repo.go]:
```suggestion
```

File: models/repo/repo.go [hunk 4]
```diff
@@ -765,35 +765,35 @@ func CountRepositories(ctx context.Context, opts CountRepositoryOptions) (int64,
...
+	var cond builder.Cond = builder.Eq{
+		"repo_id": repoID,
+		"is_pull": isPull,
+	}
+	if isClosed {
+		cond = cond.And(builder.Eq{"is_closed": isClosed})
+	}
+	subQuery := builder.Select("count(*)").From("issue").Where(cond)
+
+	query := builder.Update(builder.Eq{field: subQuery}).
```
Comment (delvh) [inline models/repo/repo.go:787]: Yeah, I can absolutely understand why the syntax highlighter is absolutely confused.
Got me as well for a minute.

File: models/repo/repo.go [hunk 5]
```diff
@@ -765,35 +765,35 @@ func CountRepositories(ctx context.Context, opts CountRepositoryOptions) (int64,
...
 }
 
-// StatsCorrectNumClosed update repository's issue related numbers
-func StatsCorrectNumClosed(ctx context.Context, id int64, isPull bool, field string) error {
-	_, err := db.Exec(ctx, "UPDATE `repository` SET "+field+"=(SELECT COUNT(*) FROM `issue` WHERE repo_id=? AND is_closed=? AND is_pull=?) WHERE id=?", id, true, isPull, id)
-	return err
-}
-
-// UpdateRepoIssueNumbers update repository issue numbers
+// UpdateRepoIssueNumbers update repository's issue closed numbers
```
Comment (delvh) [inline models/repo/repo.go]:
```suggestion
// UpdateRepoIssueNumbers updates one of a repositories amount of (open|closed) (issues|PRs) with the current count 
```

File: models/repo/repo.go (hunk 3 above)
Comment (delvh) [reply models/repo/repo.go]: For the record: I had to split up this suggestion into three comments as GitHub seems to have trouble defining their requirements for when you can add a suggestion.

**Comment by lunny** [comment]
@delvh done.

**Comment by lunny** [comment]
> 

It should not be fixed. You have no more tests so that you cannot encounter it.

**Review by delvh — APPROVED** [review]

File: models/repo/repo.go [hunk 6]
```diff
@@ -765,35 +765,32 @@ func CountRepositories(ctx context.Context, opts CountRepositoryOptions) (int64,
...
+	var cond builder.Cond = builder.Eq{
+		"repo_id": repoID,
+		"is_pull": isPull,
+	}
+	if isClosed {
+		cond = cond.And(builder.Eq{"is_closed": isClosed})
+	}
+	subQuery := builder.Select("count(*)").From("issue").Where(cond)
+
+	query := builder.Update(builder.Eq{field: subQuery}).
```
Comment (6543) [inline models/repo/repo.go:787]: is this the intended usage? to use builder eq for an update statement?!?

Comment (delvh) [reply models/repo/repo.go:787]: What is going on here is that the builder is misused to set the field to update dynamically.
But yeah, maybe we should document this line as I've also struggled with it, and it certainly isn't as apparent as is typical for Go code.

Comment (6543) [reply models/repo/repo.go:787]: for my understanding `builder.Eq{field: subQuery}` is equivalten to go-code `fmt.Sprintf("%s = %s", field, subQuery)` but better sanitized

Comment (lunny) [reply models/repo/repo.go:787]: Here Update means Set and Update

Comment (6543) [reply models/repo/repo.go:787]: @lunny can you add some code comment to explain for others :)

Comment (lunny) [reply models/repo/repo.go:787]: Done. And in fact, this is not the first place we use `builder.Update` in codebase.

**Review by 6543 — APPROVED** [review]

**Pull request merged**